---
title: "Example for spSeudoMap in R (Simulation data)"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{spSeudoMap_simulation}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)

knitr::opts_knit$set(
  root.dir = rprojroot::find_rstudio_root_file()
)
```

## Install required packages
```{r setup}
# ## Install devtools
# if (!requireNamespace("devtools", quietly = TRUE))
#   install.packages('devtools')
# 
# ## Install spSeudoMap package
# if (!requireNamespace("spSeudoMap", quietly = TRUE))
#   devtools::install_github("bsungwoo/spSeudoMap")
# 
# ## Install sceasy
# # https://github.com/cellgeni/sceasy
# # 1. Install BiocManager
# if (!requireNamespace("BiocManager", quietly = TRUE))
#   install.packages("BiocManager")
# 
# # 2. Install LoomExperiment
# if (!requireNamespace("LoomExperiment", quietly = TRUE))
#   BiocManager::install(c("LoomExperiment", "SingleCellExperiment"))
# 
# # 3. Install sceasy
# if (!requireNamespace("sceasy", quietly = TRUE))
#   devtools::install_github("cellgeni/sceasy")

```

## Load required packages
```{r}
# library(spSeudoMap)
# library(SeuratObject)
```


## Make output and data directories
```{r}
# # Set the location of the R project file as a default directory
# filepath <- rprojroot::find_rstudio_root_file()
# # Set working directoy
# setwd(filepath)
# 
# # Create output directory
# output_folder_name <- 'output'
# if (!file.exists(output_folder_name)){
#   dir.create(output_folder_name)
# }
# 
# # Create data directory
# if (!file.exists('data')){
#   dir.create('data')
# }
```


## Install conda environment
<!-- conda create -n spSeudoMap python=3.7.12 -c conda-forge -->
<!-- conda activate spSeudoMap -->
<!-- pip install git+https://github.com/bsungwoo/spSeudoMap.git -->
<!-- python -m ipykernel install --user --name spSeudoMap --display-name spSeudoMap -->


## Load single-cell and spatial datasets
### Download single-nucleus dataset
#### Kleshchevnikov, V., Shmatko, A., Dann, E. et al. Cell2location maps fine-grained cell types in spatial transcriptomics. Nat Biotechnol (2022).
#### https://cell2location.readthedocs.io/en/latest/notebooks/cell2location_short_demo.htmlÂ¶
```{r}
# options(timeout=1000)
# download.file(url = "https://cell2location.cog.sanger.ac.uk/tutorial/mouse_brain_snrna/regression_model/RegressionGeneBackgroundCoverageTorch_65covariates_40532cells_12819genes/sc.h5ad",
#               destfile = './data/adata_single.h5ad')
```


### Load single-nucleus data and generate simulation dataset
#### Load conda environment and load .h5ad
```{r}
# ## Load conda environment and convert to Seurat object
# reticulate::use_condaenv("spSeudoMap", required=TRUE)
# sc <- reticulate::import('scanpy', convert = FALSE)
# sp <- reticulate::import('scipy.sparse', convert = FALSE)
# 
# # Load single-cell data 
# sc_data <- sc$read(file.path(filepath, './data/adata_single.h5ad'))

```

#### Extract raw data, convert to the matrix and extract metadata
```{r}
# # Extract raw data
# sc_data = sc_data$raw$to_adata()
# 
# # Convert scipy matrix to sparse matrix in R
# X <- Matrix::t(reticulate::py_to_r(sp$csc_matrix(sc_data$X)))
# # Assign row names and column names
# rownames(X) <- reticulate::py_to_r(sc_data$var['SYMBOL']$tolist())
# colnames(X) <- reticulate::py_to_r(sc_data$obs$index$tolist())
# 
# # Convert metadata(.obs) in pandas dataframe into R dataframe
# obs <- reticulate::py_to_r(sc_data$obs)
```

#### Create Seurat Object and save
```{r}
# # Convert to Seurat object and add metadata
# sc_data_r <- Seurat::CreateSeuratObject(X)
# sc_data_r <- Seurat::AddMetaData(sc_data_r, obs)
```

#### (Load the Seurat object) and create simulation subpopulation data (with Ext neurons)
```{r}
# # Subset excitatory neurons
# ext_neuron_types <- grep(pattern='^Ext_', 
#                          levels(factor(sc_data_r[['annotation_1']][,1])), value=T)
# ext_neuron_types
# 
# # Set the identity to the cell types
# Seurat::Idents(sc_data_r) <- sc_data_r[['annotation_1']]
# 
# # Subset the single-cell data to contain only the excitatory neurons
# sc_data_sim <- subset(sc_data_r, idents=ext_neuron_types)
# 
# # Save the subsetted simulated single-cell data (Ext)
# saveRDS(sc_data_sim, file.path(filepath,'./data/sc_data_sim.rds'))
```


### Load spatial dataset (RDS file with Seurat object): 10X genomics data repository
#### V1_Adult_Mouse_Brain_Coronal_Section_1
```{r}
# # Import dataset from scanpy
# sc$datasets$visium_sge(
#   sample_id="V1_Adult_Mouse_Brain_Coronal_Section_1"
# 
# )
# 
# sp_data <- Seurat::Load10X_Spatial(file.path(filepath,'./data/V1_Adult_Mouse_Brain_Coronal_Section_1'))
```

### Check the size of spatial data
```{r}
# dim(sp_data)
```

### Set the number of pseudospots: 5 times the number of spatial spots
```{r}
# npseudo <- 5*dim(sp_data)[2]
# npseudo
```

### Set the pseudotype fraction from the simulation dataset (fraction of non-excitatory neuron population)
```{r}
# # Calculate pseudotype fraction from the simulation dataset
# pseudo_frac_m <- 1 - ((dim(sc_data_sim)[2])/(dim(sc_data_r)[2]))
# pseudo_frac_m
```

## Perform spSeuoMap analysis
#### Explanation for the function
```{r}
# help("pred_cellf_spSeudoMap")
```

### Run spSeudoMap
```{r}
# # Using already installed conda environment
# # Conda environment will be automatically installed if 'spSeudoMap' is not already installed
# sp_data_cellf <- pred_cellf_spSeudoMap(sp_data,sc_data_sim,
#                                        outdir=output_folder_name,
#                                        sp_subset=F,spot.cluster.name='seurat_clusters',
#                                        spot.cluster.of.interest=NULL,
#                                        env.select='conda',python.install=T,
#                                        python_path=NULL,env.name='spSeudoMap',
#                                        gpu=TRUE,metadata_celltype='annotation_1',
#                                        num_markers=40,mixture_mode='pseudotype',
#                                        seed_num=0,
#                                        mk_ratio_fix=T, mk_ratio=4,
#                                        pseudo_frac_m=pseudo_frac_m, pseudo_frac_std=0.1, 
#                                        nmix=8,npseudo=npseudo,alpha=0.6,alpha_lr=5,
#                                        emb_dim=64,batch_size=512,n_iterations=3000,init_train_epoch=10)

```


### Save seurat object with cell fraction
```{r}
# saveRDS(sp_data_cellf, file.path(filepath, output_folder_name, 'sp_data_cellf.rds'))
```



### Visualization of spatial cell fraction
#### Remove '_cellf' from the column names cell fraction metadata
```{r}
# cellf.data <- sp_data_cellf@meta.data
# cellf.data.colname <- sapply(colnames(cellf.data), function(x){
#   if (grepl('_cellf',x)){return(strsplit(x, split='_cellf')[[1]][1])}
#   else {return(x)}
# })
# sp_data_cellf.mod <- sp_data_cellf
# colnames(sp_data_cellf.mod@meta.data) <- cellf.data.colname
```


#### Visualize the layer-specific excitatory neuons
```{r}
# cell_types <- c('Ext_L25','Ext_L23','Ext_L56','Ext_L5_1',
#                 'Ext_L6B','Ext_Hpc_CA1','Ext_Hpc_CA3',
#                 'Ext_Hpc_DG1','Ext_Thal_1','Ext_Pir','Ext_Amy_1','Others')
# 
# p <- Seurat::SpatialFeaturePlot(sp_data_cellf.mod, features = cell_types, 
#                                 ncol = 4, alpha=0.6, combine = FALSE)
# patchwork::wrap_plots(p, ncol = 4)

```
